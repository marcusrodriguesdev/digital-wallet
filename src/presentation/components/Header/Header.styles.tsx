import styled from 'styled-components'

export const Container = styled.header`
  align-items: center;
  border-bottom: ${({ theme }) => theme?.colors?.cardBackground} 1px solid;
  display: flex;
  justify-content: space-between;
  padding: 20px 48px;
`

export const MenuButton = styled.button`
  background: none;
  border: none;
  cursor: pointer;
`

export const ContainerInfo = styled.div`
  display: flex;
`

export const Label = styled.p`
  ${({ theme }) => theme?.fonts?.bigInfo ?? {}}
  color: ${({ theme }) => theme?.colors?.subHeadline ?? '#94a1b2'};
  display: inline;
`

interface OpenInputProps {
  value: 'expense' | 'income' | 'month'
}

export const OpenInput = styled.button<OpenInputProps>`
  background: none;
  border: none;
  cursor: pointer;
  margin-right: 36px;

  span {
    ${({ theme }) => theme?.fonts?.bigInfo ?? {}}
    color: ${({ theme, value }) =>
      value === 'expense'
        ? theme?.colors?.error ?? '#ff6666'
        : value === 'income'
        ? theme?.colors?.tertiary ?? '#2cb67d'
        : theme?.colors?.highlight ?? '#7f5af0'};
  }
`

export const CircleName = styled.div`
  align-items: center;
  background: ${({ theme }) => theme?.colors?.subHeadline ?? '#94a1b2'};
  border-radius: 30px;
  display: flex;
  height: 44px;
  justify-content: center;
  width: 44px;

  span {
    ${({ theme }) => theme?.fonts?.bigInfo ?? {}}
    background: none;
    color: ${({ theme }) => theme?.colors?.cardBackground ?? '#16161a'};
    cursor: default;
  }
`
