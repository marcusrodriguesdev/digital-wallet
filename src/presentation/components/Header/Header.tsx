import React from 'react'

import { Assets } from '@/presentation/theme/Assets'
import { getInitials } from '@/presentation/utils/string'

import { CircleName, Container, ContainerInfo, OpenInput, Label, MenuButton } from './Header.styles'

export interface HeaderProps {
  renderInfo: boolean
  email: string
  expenseTotal: number
  incomeTotal: number
  onExpenseClick: () => void
  onIncomeClick: () => void
  onOpenDrawer: () => void
}

/**
 * A Header for the application
 *
 * @param renderInfo      The boolean to render the info
 * @param email           The email of the user
 * @param expenseTotal    The total of expenses
 * @param incomeTotal     The total of incomes
 * @param onExpenseClick  A function to be called when the expense button is clicked
 * @param onIncomeClick   A function to be called when the income button is clicked
 * @param onOpenDrawer    A function to be called when the menu button is clicked
 */

const Header: React.FC<HeaderProps> = (props: HeaderProps) => {
  const { email, renderInfo, expenseTotal, incomeTotal, onExpenseClick, onIncomeClick, onOpenDrawer } = props

  return (
    <Container>
      <MenuButton onClick={onOpenDrawer} data-testid="Header/MenuButton">
        <img src={Assets.HamburguerMenu} alt="menu" />
      </MenuButton>
      <ContainerInfo>
        {renderInfo && (
          <>
            <OpenInput expense onClick={onExpenseClick} data-testid="Header/ExpenseButton">
              <Label>Total de despesas: </Label>
              <span>R${expenseTotal}</span>
            </OpenInput>
            <OpenInput expense={false} onClick={onIncomeClick} data-testid="Header/IncomeButton">
              <Label>Total de ganhos: </Label>
              <span>R${incomeTotal}</span>
            </OpenInput>
          </>
        )}
        <CircleName data-testid="Header/CircleName">
          <span>{getInitials(email)}</span>
        </CircleName>
      </ContainerInfo>
    </Container>
  )
}

export { Header }
