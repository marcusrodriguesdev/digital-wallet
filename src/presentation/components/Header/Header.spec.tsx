/* eslint-disable @typescript-eslint/restrict-template-expressions */
import React from 'react'

import { fireEvent, render } from '@testing-library/react'
import * as Faker from 'faker'

import { Header, HeaderProps } from './Header'

const makeSut = (props?: Partial<HeaderProps>): any => {
  const expenseTotal = props?.expenseTotal ?? Faker.random.number()
  const incomeTotal = props?.incomeTotal ?? Faker.random.number()
  const email = props?.email ?? Faker.internet.email()
  const month = props?.month ?? Faker.date.month()
  const renderInfo = props?.renderInfo ?? true
  const onExpenseClick = jest.fn()
  const onIncomeClick = jest.fn()
  const onOpenDrawer = jest.fn()
  const onDateClick = jest.fn()

  const sut = (
    <Header
      month={month}
      renderInfo={renderInfo}
      email={email}
      expenseTotal={expenseTotal}
      incomeTotal={incomeTotal}
      onExpenseClick={onExpenseClick}
      onIncomeClick={onIncomeClick}
      onOpenDrawer={onOpenDrawer}
      onDateClick={onDateClick}
    />
  )

  return {
    sut,
    renderInfo,
    email,
    expenseTotal,
    incomeTotal,
    onExpenseClick,
    onIncomeClick,
    onOpenDrawer,
    onDateClick
  }
}

describe('presentation/components/Header', () => {
  it('should match snapshot', () => {
    const { sut } = makeSut({
      email: 'alguem@gmail.com',
      expenseTotal: 100,
      incomeTotal: 200,
      month: 'Janeiro'
    })
    const component = render(sut)
    expect(component).toMatchSnapshot()
  })

  it('should call onExpenseClick when expense button is clicked', () => {
    const { sut, onExpenseClick } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/ExpenseButton')

    fireEvent.click(button)

    expect(onExpenseClick).toBeCalledTimes(1)
  })

  it('should call onIncomeClick when income button is clicked', () => {
    const { sut, onIncomeClick } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/IncomeButton')

    fireEvent.click(button)

    expect(onIncomeClick).toBeCalledTimes(1)
  })

  it('should call onOpenDrawer when menu button is clicked', () => {
    const { sut, onOpenDrawer } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/MenuButton')

    fireEvent.click(button)

    expect(onOpenDrawer).toBeCalledTimes(1)
  })

  it('should display expense total', () => {
    const { sut, expenseTotal } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/ExpenseButton')

    expect(button.textContent).toContain(`total de despesas: R$${expenseTotal}`)
  })

  it('should display income total', () => {
    const { sut, incomeTotal } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/IncomeButton')

    expect(button.textContent).toContain(`total de ganhos: R$${incomeTotal}`)
  })

  it('should display initials', () => {
    const { sut } = makeSut({
      email: 'alguem@gmail.com'
    })

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/CircleName')

    expect(button.textContent).toContain('A')
  })

  it('should call onDateClick when date button is clicked', () => {
    const { sut, onDateClick } = makeSut()

    const { getByTestId } = render(sut)
    const button = getByTestId('Header/DateButton')

    fireEvent.click(button)

    expect(onDateClick).toBeCalledTimes(1)
  })
})
