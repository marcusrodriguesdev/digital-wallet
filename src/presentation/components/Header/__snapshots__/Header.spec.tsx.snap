// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presentation/components/Header should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="sc-bcXHqe eHDFxN"
      >
        <button
          class="sc-gswNZR bVtgCF"
          data-testid="Header/MenuButton"
        >
          <img
            alt="menu"
            src="[object Object]"
          />
        </button>
        <div
          class="sc-dkrFOg kJEZRB"
        >
          <button
            class="sc-eDvSVe iUasUO"
            data-testid="Header/DateButton"
            value="month"
          >
            <p
              class="sc-hLBbgP kCFwqV"
            >
              mês: 
            </p>
            <span>
              Janeiro
            </span>
          </button>
          <button
            class="sc-eDvSVe eKDymJ"
            data-testid="Header/ExpenseButton"
            value="expense"
          >
            <p
              class="sc-hLBbgP kCFwqV"
            >
              total de despesas: 
            </p>
            <span>
              R$
              100
            </span>
          </button>
          <button
            class="sc-eDvSVe yYooT"
            data-testid="Header/IncomeButton"
            value="income"
          >
            <p
              class="sc-hLBbgP kCFwqV"
            >
              total de ganhos: 
            </p>
            <span>
              R$
              200
            </span>
          </button>
          <div
            class="sc-jSUZER brNRCF"
            data-testid="Header/CircleName"
          >
            <span>
              A
            </span>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="sc-bcXHqe eHDFxN"
    >
      <button
        class="sc-gswNZR bVtgCF"
        data-testid="Header/MenuButton"
      >
        <img
          alt="menu"
          src="[object Object]"
        />
      </button>
      <div
        class="sc-dkrFOg kJEZRB"
      >
        <button
          class="sc-eDvSVe iUasUO"
          data-testid="Header/DateButton"
          value="month"
        >
          <p
            class="sc-hLBbgP kCFwqV"
          >
            mês: 
          </p>
          <span>
            Janeiro
          </span>
        </button>
        <button
          class="sc-eDvSVe eKDymJ"
          data-testid="Header/ExpenseButton"
          value="expense"
        >
          <p
            class="sc-hLBbgP kCFwqV"
          >
            total de despesas: 
          </p>
          <span>
            R$
            100
          </span>
        </button>
        <button
          class="sc-eDvSVe yYooT"
          data-testid="Header/IncomeButton"
          value="income"
        >
          <p
            class="sc-hLBbgP kCFwqV"
          >
            total de ganhos: 
          </p>
          <span>
            R$
            200
          </span>
        </button>
        <div
          class="sc-jSUZER brNRCF"
          data-testid="Header/CircleName"
        >
          <span>
            A
          </span>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
